src/apps/external_executor/external_executor.cpp

```diff
@@ -129,6 +124,147 @@ namespace externalexecutor
+            strategy,
+            data_structure,
+            executor_id,
+            ctx,
+            node_context,
+            std::move(out_stream));
+
+        node_context.get_indexing_strategies().install_strategy(map_index);
+
+        // store the index pointer into the strategies
+        map_index_strategies[strategy] = map_index;
+
+        return ccf::grpc::make_pending();
+      };
+
+      make_endpoint(
+        "/externalexecutor.protobuf.Index/InstallAndSubscribe",
+        HTTP_POST,
+        ccf::grpc_command_unary_stream_adapter<
+          externalexecutor::protobuf::IndexInstall,
+          externalexecutor::protobuf::IndexWork>(install_and_subscribe_index),
+        executor_only)
+        .set_forwarding_required(ccf::endpoints::ForwardingRequired::Never)
+        .install();
+
+      auto store_indexed_data =
+        [this](
+          ccf::endpoints::EndpointContext& ctx,
+          externalexecutor::protobuf::IndexPayload&& payload)
+        -> ccf::grpc::GrpcAdapterResponse<google::protobuf::Empty> {
+        std::string strategy = payload.strategy_name();
+        auto it = map_index_strategies.find(strategy);
+        if (it == map_index_strategies.end())
+        {
+          return ccf::grpc::make_error(
+            GRPC_STATUS_NOT_FOUND,
+            fmt::format("Strategy {} doesn't exist", strategy));
+        }
+
+        std::string key = payload.key();
+        std::string val = payload.value();
+        map_index_strategies[strategy]->store(key, val);
+        return ccf::grpc::make_success();
+      };
+
+      make_endpoint(
+        "/externalexecutor.protobuf.Index/StoreIndexedData ",
+        HTTP_POST,
+        ccf::grpc_adapter<
+          externalexecutor::protobuf::IndexPayload,
+          google::protobuf::Empty>(store_indexed_data),
+        executor_only)
+        .set_forwarding_required(ccf::endpoints::ForwardingRequired::Never)
+        .install();
+
+      auto get_indexed_data = [this](
+                                ccf::endpoints::ReadOnlyEndpointContext& ctx,
+                                externalexecutor::protobuf::IndexKey&& payload)
+        -> ccf::grpc::GrpcAdapterResponse<
+          externalexecutor::protobuf::OptionalKVValue> {
+        std::string strategy_name = payload.strategy_name();
+
+        auto it = map_index_strategies.find(strategy_name);
+        if (it == map_index_strategies.end())
+        {
+          return ccf::grpc::make_error(
+            GRPC_STATUS_NOT_FOUND,
+            fmt::format("Index {} is not found", strategy_name));
+        }
+        auto executor_id = get_caller_executor_id(ctx);
+        auto strategy_ptr = map_index_strategies[strategy_name];
+        std::string key = payload.key();
+        std::optional<std::string> data = strategy_ptr->fetch(key);
+
+        if (!data.has_value())
+        {
+          return ccf::grpc::make_error(
+            GRPC_STATUS_NOT_FOUND, "Key was not found in the indexed data");
+        }
```

Comment:
This path will be fairly common, and I'd lean towards implementing as an empty response (ie - `value` has no body => key wasn't found), rather than an error type. This will make it simpler to extend with different kinds of error in future (eg - distinguish "I've never been told this" from "I think I have, but have to fetch it asynchronously from disk"), handled inline without exception handling.

Reply:
Agree, modified this